### ISSUE TITLE
JSONDecodeError

### ISSUE THREAD

**Issue by pigmonkey** [alpernebbi/git-annex-metadata-gui#1]
I'm getting a couple JSON decode errors when I try to use this. On one annex, I get an error immediately upon execution:

	$ git-annex-metadata-gui ~/library/pictures
	Traceback (most recent call last):
	  File "/home/pigmonkey/.virtualenvs/annexgui/bin/git-annex-metadata-gui", line 11, in <module>
	    load_entry_point('git-annex-metadata-gui==0.1.0', 'gui_scripts', 'git-annex-metadata-gui')()
	  File "/home/pigmonkey/.virtualenvs/annexgui/lib/python3.5/site-packages/git_annex_metadata_gui/gui.py", line 43, in main
	    window.load_repository(sys.argv[1])
	  File "/home/pigmonkey/.virtualenvs/annexgui/lib/python3.5/site-packages/git_annex_metadata_gui/gui.py", line 157, in load_repository
	    self.models.keys = GitAnnexKeysModel(self.annex)
	  File "/home/pigmonkey/.virtualenvs/annexgui/lib/python3.5/site-packages/git_annex_metadata_gui/models.py", line 41, in __init__
	    for key in self.annex.keys(absent=True, cached=True)
	  File "/home/pigmonkey/.virtualenvs/annexgui/lib/python3.5/site-packages/git_annex_adapter.py", line 234, in keys
	    all_meta = self.metadata(all_keys=True, cached=cached)
	  File "/home/pigmonkey/.virtualenvs/annexgui/lib/python3.5/site-packages/git_annex_adapter.py", line 226, in metadata
	    metadata = [json.loads(json_) for json_ in jsons]
	  File "/home/pigmonkey/.virtualenvs/annexgui/lib/python3.5/site-packages/git_annex_adapter.py", line 226, in <listcomp>
	    metadata = [json.loads(json_) for json_ in jsons]
	  File "/usr/lib64/python3.5/json/__init__.py", line 319, in loads
	    return _default_decoder.decode(s)
	  File "/usr/lib64/python3.5/json/decoder.py", line 339, in decode
	    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
	  File "/usr/lib64/python3.5/json/decoder.py", line 355, in raw_decode
	    obj, end = self.scan_once(s, idx)
	json.decoder.JSONDecodeError: Expecting property name enclosed in double quotes: line 1 column 118 (char 117)

On another annex, I can succesfully load the GUI, but as soon as I click a file (any file), it crashes with an error:

	$ git-annex-metadata-gui ~/library/tracks
	Traceback (most recent call last):
	  File "/home/pigmonkey/.virtualenvs/annexgui/lib/python3.5/site-packages/git_annex_metadata_gui/gui.py", line 223, inselection_updated
	    self.docks.editor.set_item(item)
	  File "/home/pigmonkey/.virtualenvs/annexgui/lib/python3.5/site-packages/git_annex_metadata_gui/docks.py", line 145, in set_item
	    for field in item:
	  File "/home/pigmonkey/.virtualenvs/annexgui/lib/python3.5/site-packages/git_annex_adapter.py", line 341, in __iter__
	    for field in self.fields().keys():
	  File "/home/pigmonkey/.virtualenvs/annexgui/lib/python3.5/site-packages/git_annex_adapter.py", line 294, in fields
	    new_fields = self.query(fields=fields).get('fields', {})
	  File "/home/pigmonkey/.virtualenvs/annexgui/lib/python3.5/site-packages/git_annex_adapter.py", line 413, in __call__
	    return response if query_line else json.loads(response)
	  File "/usr/lib64/python3.5/json/__init__.py", line 319, in loads
	    return _default_decoder.decode(s)
	  File "/usr/lib64/python3.5/json/decoder.py", line 339, in decode
	    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
	  File "/usr/lib64/python3.5/json/decoder.py", line 357, in raw_decode
	    raise JSONDecodeError("Expecting value", s, err.value) from None
	json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
	[1]    5572 abort (core dumped)  git-annex-metadata-gui ~/library/tracks

Here is some version info for my virtualenv:

	$ pip freeze
	git-annex-adapter==0.1.0
	git-annex-metadata-gui==0.1.0
	PyQt5==5.7
	sip==4.18.1
	$ python --version
	Python 3.5.2

Let me know if there is something I can do to provide more useful debugging info.

**Comment by alpernebbi** [comment]
Looks like the first error happens while decoding the output of `git-annex metadata --all --json`. Extract all metadata (in `~/library/pictures`),
```
$ git-annex metadata --all --json > all_meta.txt
```
Then check which line gives a decode error:
```
#!/usr/bin/env python3
import json

with open('all_meta.txt') as f:
    lines = f.readlines()

for line in lines:
    try:
        json.loads(line)
    except json.JSONDecodeError as e:
        print(line, e, sep='\n', end='\n\n')
```
I guess most of the problematic lines will have the same root cause, so looking at one or two would be enough.

In the second one, the reply to `{"key":"...","fields":null}` from `git-annex metadata --batch --json` seems to be empty (i.e. `json.loads('')`). Take the key of a file that crashes the GUI, and check that the problem isn't in git-annex: 
```
$ key=$(git-annex lookupkey file.ext)
$ git-annex metadata --key $key
$ echo {"key":"'$key'"} | git-annex metadata --batch --json
$ echo {"key":"'$key'","fields":null} | git-annex metadata --batch --json
```
Also check if the outputs of the last two give errors when you `json.loads()` them in python. If the files that crash the GUI don't give errors here, apply the following patch to `git_annex_adapter.py`:
```
diff --git a/git_annex_adapter.py b/git_annex_adapter.py
index c08c214..880159f 100644
--- a/git_annex_adapter.py
+++ b/git_annex_adapter.py
@@ -410,7 +410,11 @@ class BatchProcess:
         query = " ".join(query_line) or json.dumps(query_object)
         print(query, file=self._process.stdin, flush=True)
         response = self._process.stdout.readline().strip()
-        return response if query_line else json.loads(response)
+        try:
+            return response if query_line else json.loads(response)
+        except:
+            print(query, response, sep='\n', end='\n\n')
+            raise
 
     def __repr__(self):
         repr_ = 'BatchProcess(cmd={!r}, cwd={!r}, process={!r})'
```
Install it and test again. The printed query should be in the form of `{"key":"...","fields":null}`, but the response should be empty. Try this query in `$ echo '{...}' | git-annex metadata --batch --json` as well.

As a result of these, you will probably isolate a key (with fields) which triggers these errors. I would appreciate it if you could replicate these conditions in new repositories, compress (`tar.gz`) and attach them here; but it's also OK if you just tell me the conditions.

**Comment by pigmonkey** [comment]
For the first one, it looks like I have some files in that annex which have no metadata.

	{"command":"metadata","key":"SHA256E-s1458084--333588588d6bf5463d3095fdfc89504a93fedd048e049b983b1aeb0943b91d22.jpg",,"note":"","success":true}

	Expecting property name enclosed in double quotes: line 1 column 118 (char 117)

	{"command":"metadata","key":"SHA256-s4435456--f601445239da3057d376f3bbd9d5c83871d1545b12bf2d4409096ebea66535c7",,"note":"","success":true}

	Expecting property name enclosed in double quotes: line 1 column 113 (char 112)

	{"command":"metadata","key":"SHA256-s2568433--e68cebe40e379cdb22f5760af4b1dafb647bc50d757347aaac686fba7112d3b8",,"note":"","success":true}

	Expecting property name enclosed in double quotes: line 1 column 113 (char 112)

	{"command":"metadata","key":"SHA256E-s1311446--b776c32cd96b6a0e4e044badebe30528321e397668e7104121c322ac18c59502.jpg",,"note":"","success":true}

	Expecting property name enclosed in double quotes: line 1 column 118 (char 117)

	{"command":"metadata","key":"SHA256-s115259--26476281fc1750c68357cb74d78c0e00d756d95a7c21ad81814e211eea031858",,"note":"","success":true}

	Expecting property name enclosed in double quotes: line 1 column 112 (char 111)

	{"command":"metadata","key":"SHA256-s2106181--c9ea9750218747da1b963048b09b2c2e74ca0bf2fc991ceca8f43faa14646dd4",,"note":"","success":true}

I recursively set a tag of `photo` in the annex so that they all have something. Oddly it seems there are still keys in the annex without any metadata.

	$ git-annex metadata --force -t photo
	...
	$ git-annex metadata --all --json > ~/all_meta.txt
	$ head ~/all_meta.txt
	{"command":"metadata","key":"SHA256E-s1458084--333588588d6bf5463d3095fdfc89504a93fedd048e049b983b1aeb0943b91d22.jpg","lastchanged":["2016-12-08@17-27-00"],"tag":["photo"],"tag-lastchanged":["2016-12-08@17-27-00"],"note":"lastchanged=2016-12-08@17-27-00\ntag=photo\ntag-lastchanged=2016-12-08@17-27-00\n","success":true}
	{"command":"metadata","key":"SHA256-s4435456--f601445239da3057d376f3bbd9d5c83871d1545b12bf2d4409096ebea66535c7",,"note":"","success":true}
	{"command":"metadata","key":"SHA256-s2568433--e68cebe40e379cdb22f5760af4b1dafb647bc50d757347aaac686fba7112d3b8",,"note":"","success":true}
	{"command":"metadata","key":"SHA256E-s1311446--b776c32cd96b6a0e4e044badebe30528321e397668e7104121c322ac18c59502.jpg","lastchanged":["2016-12-08@17-27-00"],"tag":["photo"],"tag-lastchanged":["2016-12-08@17-27-00"],"note":"lastchanged=2016-12-08@17-27-00\ntag=photo\ntag-lastchanged=2016-12-08@17-27-00\n","success":true}
	{"command":"metadata","key":"SHA256-s115259--26476281fc1750c68357cb74d78c0e00d756d95a7c21ad81814e211eea031858",,"note":"","success":true}
	{"command":"metadata","key":"SHA256-s2106181--c9ea9750218747da1b963048b09b2c2e74ca0bf2fc991ceca8f43faa14646dd4",,"note":"","success":true}
	{"command":"metadata","key":"SHA256E-s3369088--16750b92d65a39d43d1aba6f7d51c2515bb3cb03c6b524dce67289c1db3a600a.jpg","lastchanged":["2016-12-08@17-27-00"],"tag":["photo"],"tag-lastchanged":["2016-12-08@17-27-00"],"note":"lastchanged=2016-12-08@17-27-00\ntag=photo\ntag-lastchanged=2016-12-08@17-27-00\n","success":true}
	{"command":"metadata","key":"SHA256-s3404800--7a2249d8fad1c070798085b36e9b9c569d016649f1df455a4eed064ead89173d",,"note":"","success":true}
	{"command":"metadata","key":"SHA256-s1647983--aff9935b82ad3f68c98e39f300fbd1aa9ddf42dde48b3da714e002e913a226de",,"note":"","success":true}
	{"command":"metadata","key":"SHA256E-s4377646--a6f8fdc164a598673e8923bb136970c4e3abde200bd54c1a01005eec4cb912bb.jpg","lastchanged":["2016-12-08@17-27-00"],"tag":["photo"],"tag-lastchanged":["2016-12-08@17-27-00"],"note":"lastchanged=2016-12-08@17-27-00\ntag=photo\ntag-lastchanged=2016-12-08@17-27-00\n","success":true}
	$ python test.py
	{"command":"metadata","key":"SHA256-s4435456--f601445239da3057d376f3bbd9d5c83871d1545b12bf2d4409096ebea66535c7",,"note":"","success":true}

	Expecting property name enclosed in double quotes: line 1 column 113 (char 112)

	{"command":"metadata","key":"SHA256-s2568433--e68cebe40e379cdb22f5760af4b1dafb647bc50d757347aaac686fba7112d3b8",,"note":"","success":true}

	Expecting property name enclosed in double quotes: line 1 column 113 (char 112)

	{"command":"metadata","key":"SHA256-s115259--26476281fc1750c68357cb74d78c0e00d756d95a7c21ad81814e211eea031858",,"note":"","success":true}

	Expecting property name enclosed in double quotes: line 1 column 112 (char 111)

	{"command":"metadata","key":"SHA256-s2106181--c9ea9750218747da1b963048b09b2c2e74ca0bf2fc991ceca8f43faa14646dd4",,"note":"","success":true}

	Expecting property name enclosed in double quotes: line 1 column 113 (char 112)

	{"command":"metadata","key":"SHA256-s3404800--7a2249d8fad1c070798085b36e9b9c569d016649f1df455a4eed064ead89173d",,"note":"","success":true}

	Expecting property name enclosed in double quotes: line 1 column 113 (char 112)
	...

I'm not sure what those other keys are (I checked `unused`, but there is nothing there). This particular annex was created in 2012, so there could be a lot of cruft in there. Anyway, it seems like we would want the GUI to just pass on files where it cannot find anything useful.

I'll start another issue for the second repo so that we can keep them separate.

**Comment by alpernebbi** [comment]
From `6.20160726` on, git-annex uses a `fields` object to wrap the metadata fields in the JSON output. Right now, I rely on that format, but your git-annex version uses the old one.

I'm keeping track of this in alpernebbi/git-annex-adapter#1, but keeping this open until I fix that.

**Comment by pigmonkey** [comment]
This is resolved now that Arch is on 6.20161118.
